### ISSUE TITLE
Possibility to produce heterogeneous datasets

### ISSUE THREAD

**Issue by drudgede** [vincentrussell/json-data-generator#25]
Hello there,

is it possible to produce heterogeneous datasets (with optional properties), e.g. by specifying the probability if a property is generated?
I tried to use `'{{repeat(0,1)}}',` to specify that a certain (property) set is generated or not. However, this did not work for me.
I mitigated it by creating properties with generator functions that have a certain probability to return an empty string instead of an actual value. However, my desire is not to have a property at all sometimes.

Is there a possibility to archive this yet?

**Comment by vincentrussell** [comment]
Can you provide an example of how you would expect the syntax to look and what the expected json would look like?  I want to make sure I understand what you are asking.  It sound like you would like a repeat to work 0 or n times based on some randomness, but I am not sure.

**Comment by vincentrussell** [comment]
Does this address what you meant?

509c16e

**Comment by drudgede** [comment]
Hello Vincent,

thanks for the commit which makes repeats inclusive. I will try out if this works for single properties, too.
My general desire was the following idea: Imagine I want to create the following minimal JSON documents:

`
    {
        "id": "1",
        "firstname": "John",
    },
    {
        "id": "1",
        "firstname": "Max",
        "isAdmin": true
    },
`

Here, I want to create the property `"isAdmin"` with a given probability, e.g. 1% of all documents should contain it. For the rest of the documents, I do not want to have the property with the value `"isAdmin": false" but I want to skip this property in the generation process. In general, I want to be able to generate heterogeneous documents as this is typical for JSON.

I don't know exactly the internals of your generation process but for me, a potential, desired solution would look like the following:

I create a custom generator class, e.g. an `isAdminGenerator` and pass the probability of the property to be generated as a parameter to the function which is annotated with `@FunctionInvocation`. An example skeleton of the function looks like:

`
    @FunctionInvocation
    public String generateIsAdminProperty(float generationProbability) {
        boolean b;
       
        // application logic that might sets "b" to true

        return b;
    }
`

And in the schema input file, it would look like the following:

`
[
    '{{repeat(2)}}',
    {
        "id": "1",
        "firstname": "{{firstName()}}",
        "isAdmin": "{{generateIsAdminProperty(0.01)}}"
    }
]
`

A potential solution that came into my mind is now the following: Is it possible to let `generateIsAdminProperty` return e.g. a `null` value to tell the generator/parser that this property should be omitted for the currently generated document?
Currently, wenn I return `null` in the function, I get a stack trace since the generator expects a string. However, is it possible to make a `null` return value a valid, distinguished value to denote that the property shall not be generated?

Or are there other possible solutions?

**Comment by vincentrussell** [comment]
having a property either be present or not present is definitely not possible as things stand now, but couldn't you just do something like this where it would just be false when not true?:

[
    '{{repeat(2)}}',
    {
        "id": "1",
        "firstname": "{{firstName()}}",
        "isAdmin": {{bool(0.1)}}
    }
]

**Comment by drudgede** [comment]
Thanks for the clarification that this is yet not possible. I have already suspected that.

Well, in this case this might work with your approach (at least from the semantic point of view). But since I currently investigate the capability of various systems to deal with heterogeneous JSON files (generated by a certain, given schema), this is not a solution for me.

Nevertheless, I think I will just insert a special property value and strip out the whole line in a preprocessing step.

However, thanks for keeping up with this great project!

**Comment by drudgede** [comment]
Hello Vincent,

is is possible to push your commit `509c16e` to MVNReposititory or somewhere else?
I found out that your changes may fit my need.

I found out that you can create property multiple times, e.g.

```
[
    '{{repeat(2)}}',
    {
        "id": "{{index(1)}}",
        '{{repeat(2,3)}}',"gender": "{{gender()}}",
    }
]
```

multiple times print the gender property. In your commit, you made your repeat inclusive, that means in my understanding, that `'{{repeat(0,1}}'` may be an option to print the property once or not at all, and hence provides a solution for optional property. I'd really like to try out this solution.

Thanks!

**Comment by vincentrussell** [comment]
Ok.  I'll try to cut a release tomorrow.

**Comment by vincentrussell** [comment]
I just released 1.11
